stats/stats.go

```diff
@@ -1,122 +1,309 @@
-				perSecond := counterValue / int(interval.Seconds())
+	return s, nil
+}
 
-				logrus.Infof("STATS [%s]: %d / %s (%d/s)\n", counterName, counterValue,
-					interval, perSecond)
+// validateConfig validates that the correct options are supplied when initiating the Stats service
+func validateConfig(cfg *Config) error {
+	if cfg.FlushInterval == 0 {
+		return ErrInvalidFlushInterval
+	}
 
-				if strings.HasSuffix(counterName, "relay-producer") {
-					SetPromGauge("plumber_relay_rate", perSecond)
-					IncrPromCounter("plumber_relay_total", counterValue)
-				}
+	if cfg.ServiceShutdownCtx == nil {
+		return ErrMissingShutdownCtx
+	}
 
-				// Reset it
-				counters[counterName] = 0
-			}
+	if cfg.TSStoragePath == "" {
+		return ErrMissingTSStoragePath
+	}
 
-			return nil
-		})
-	}()
+	return nil
 }
 
-// InitPrometheusMetrics sets up prometheus counters/gauges
-func InitPrometheusMetrics() {
-	prometheusMutex.Lock()
-	defer prometheusMutex.Unlock()
+// runWatchForShutdown is a goroutine that is ran when the Stats services is created. It listens for
+// a context cancellation and will stop/flush/delete all counters
+func (s *Stats) runWatchForShutdown() error {
+	<-s.ServiceShutdownCtx.Done()
 
-	prometheusGauges["plumber_relay_rate"] = promauto.NewGauge(prometheus.GaugeOpts{
-		Name: "plumber_relay_rate",
-		Help: "Current rare of messages being relayed to Batch.sh",
-	})
+	s.countersMtx.RLock()
+	counters := s.counters // make copy of map since RemoveCounter will need to acquire mutex lock each time
+	s.countersMtx.RUnlock()
 
-	prometheusCounters["plumber_relay_total"] = promauto.NewCounter(prometheus.CounterOpts{
-		Name: "plumber_relay_total",
-		Help: "Total number of events relayed to Batch.sh",
-	})
+	for _, c := range counters {
+		s.RemoveCounter(c)
+	}
 
-	prometheusCounters["plumber_read_errors"] = promauto.NewCounter(prometheus.CounterOpts{
-		Name: "plumber_read_errors",
-		Help: "Number of errors when reading messages",
-	})
+	if err := s.storage.Close(); err != nil {
+		err = errors.Wrap(err, "unable to safely shutdown stats service database")
+		s.log.Error(err)
+		return err
+	}
+
+	return nil
+}
+
+// genCounterID generates a string which is used to identify/retreive a counter from the counters map
+func genCounterID(counterType opts.Counter_Type, resourceType opts.Counter_Resource, resourceID string) string {
+	return fmt.Sprintf("%s-%s-%s", counterType, resourceType, resourceID)
+}
 
-	prometheusCounters["plumber_grpc_errors"] = promauto.NewCounter(prometheus.CounterOpts{
-		Name: "plumber_grpc_errors",
-		Help: "Number of errors when making GRPC calls",
+// AddCounter creates a new counter and launches a runFlusher() goroutine for it
+func (s *Stats) AddCounter(cfg *opts.Counter) *Counter {
+	// Check for existing counter
+	existing, _ := s.GetCounter(cfg.Type, cfg.Resource, cfg.ResourceId)
+	if existing != nil {
+		return existing
+	}
+
+	logger := logrus.WithFields(logrus.Fields{
+		"pkg":           "stats",
```

Comment:
Include the counter friendly name in logs.

Reply:
Fixed!